pkg/kube/dialer.go

```diff
@@ -0,0 +1,336 @@
+	default:
+		err = ErrInvalidPodFQDN
+		return
+	}
+
+	// service - labels[1]
+	pod.Name = labels[0]
+	return
+}
+
+// mutex must be held while calling this
+func (p *PodDialer) connectionForPod(pod types.NamespacedName) (httpstream.Connection, error) {
+	if conn, ok := p.connections[pod]; ok {
+		return conn, nil
+	}
+
+	transport, upgrader, err := spdy.RoundTripperFor(p.config)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := p.config
+	cfg.APIPath = "/api"
+	cfg.GroupVersion = &schema.GroupVersion{Version: "v1"}
+	cfg.NegotiatedSerializer = negotiatedSerializer
+	restClient, err := rest.RESTClientFor(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	req := restClient.Post().
+		Resource("pods").
+		Namespace(pod.Namespace).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, req.URL())
+	conn, protocol, err := dialer.Dial(portForwardProtocolV1Name)
+	if err != nil {
+		return nil, err
+	}
+	if protocol != portForwardProtocolV1Name {
+		return nil, fmt.Errorf("unable to negotiate protocol: client supports %q, server returned %q", portForwardProtocolV1Name, protocol)
+	}
+
+	p.connections[pod] = conn
+	return conn, nil
+}
+
+type conn struct {
+	httpstream.Stream
+	errorStream httpstream.Stream
+	network     string
+	remote      string
+
+	errCh chan error
+}
+
+func wrapConn(network, remote string, s, err httpstream.Stream) *conn {
+	c := &conn{
+		Stream:      s,
+		errorStream: err,
+		network:     network,
+		remote:      remote,
+		errCh:       make(chan error, 1),
+	}
+
+	go c.pollErrors()
+	return c
+}
+
+func (c *conn) pollErrors() {
+	data, err := io.ReadAll(c.errorStream)
+	if err != nil {
+		c.writeError(err)
+		return
+	}
+
+	if len(data) != 0 {
+		c.writeError(errors.New(string(data)))
+		return
+	}
+}
+
+func (c *conn) writeError(err error) {
+	select {
+	case c.errCh <- err:
+	default:
+	}
+}
+
+func (c *conn) checkError() error {
+	select {
+	case err := <-c.errCh:
+		return err
+	default:
+		return nil
+	}
+}
+
+var _ net.Conn = (*conn)(nil)
+
+func (c *conn) Read(data []byte) (int, error) {
+	if err := c.checkError(); err != nil {
+		return 0, err
+	}
+
+	n, err := c.Stream.Read(data)
+
+	// prioritize any sort of checks propagated on
+	// the error stream
+	if err := c.checkError(); err != nil {
+		return n, err
+	}
+	return n, err
+}
+
+func (c *conn) Write(b []byte) (int, error) {
+	if err := c.checkError(); err != nil {
+		return 0, err
+	}
+
+	n, err := c.Stream.Write(b)
+
+	// prioritize any sort of checks propagated on
+	// the error stream
+	if err := c.checkError(); err != nil {
+		return n, err
+	}
+	return n, err
+}
+
+func (c *conn) Close() error {
+	closeErr := c.Reset()
+
+	// prioritize any sort of checks propagated on
+	// the error stream
+	if err := c.checkError(); err != nil {
+		return err
+	}
+	return closeErr
+}
+
+func (c *conn) SetDeadline(t time.Time) error {
+	if conn, ok := c.Stream.(net.Conn); ok {
```

Comment:
ooc, are there ever times that `c.Stream` would have a different implementation? As a reader of this code, I'd want to know if `c.Stream` is known to definitely or definitely not implement `net.Conn` but I now have more questions. It seems important to know one way or the other as context cancellation isn't plumbed through.

Reply:
So, would you prefer to switch to an unchecked cast? It's a bit cumbersome to track down, but the underlying k8s library eventually resolves to a [`*github.com/moby/spdystream.Stream`](https://github.com/moby/spdystream/blob/77eb080123d208697674a07b74ceaf94c98bee8b/stream.go#L323-L335) which implements `net.Conn`.

Reply:
That or perhaps return an error? I think I'd lean towards a type assertion as I wouldn't want these method calls to suddenly start to do nothing silently.